src/aws.s3/amazons3_errors.bal

```diff
@@ -0,0 +1,165 @@
+public const SERVER_SIDE_ENCRYPTION_CONFIGURATION_NOT_FOUND_ERROR = "{ballerinax/aws.s3}ServerSideEncryptionConfigurationNotFoundError";
+public const SERVICE_UNAVAILABLE = "{ballerinax/aws.s3}ServiceUnavailable";
+public const SIGNATURE_DOES_NOT_MATCH = "{ballerinax/aws.s3}SignatureDoesNotMatch";
+public const SLOW_DOWN = "{ballerinax/aws.s3}SlowDown";
+public const TEMPORARY_REDIRECT = "{ballerinax/aws.s3}TemporaryRedirect";
+public const TOKEN_REFRESH_REQUIRED = "{ballerinax/aws.s3}TokenRefreshRequired";
+public const TOO_MANY_BUCKETS = "{ballerinax/aws.s3}TooManyBuckets";
+public const UNEXPECTED_CONTENT = "{ballerinax/aws.s3}UnexpectedContent";
+public const UNRESOLVABLE_GRANT_BY_EMAIL_ADDRESS = "{ballerinax/aws.s3}UnresolvableGrantByEmailAddress";
+public const USER_KEY_MUST_BE_SPECIFIED = "{ballerinax/aws.s3}UserKeyMustBeSpecified";
+
+public type BUCKET_OPERATION_ERROR_REASONS ACCESS_DENIED|ACCOUNT_PROBLEM|ALL_ACCESS_DISABLED|
+        AMBIGUOUS_GRANT_BY_EMAIL_ADDRESS|AUTHORIZATION_HEADER_MALFORMED|BAD_DIGEST|BUCKET_ALREADY_EXISTS|
+        BUCKET_ALREADY_OWNED_BY_YOU|BUCKET_NOT_EMPTY|CREDENTIALS_NOT_SUPPORTED|CROSS_LOCATION_LOGGING_PROHIBITED|
+        ENTITY_TOO_SMALL|ENTITY_TOO_LARGE|EXPIRED_TOKEN|ILLEGAL_VERSIONING_CONFIGURATION_EXCEPTION|INCOMPLETE_BODY|
+        INCORRECT_NUMBER_OF_FILES_IN_POST_REQUEST|INLINE_DATA_TOO_LARGE|INTERNAL_ERROR|INVALID_ACCESS_KEY_ID|
+        INVALID_ADDRESSING_HEADER|INVALID_ARGUMENT|INVALID_BUCKET_NAME|INVALID_BUCKET_STATE|INVALID_DIGEST|
+        INVALID_ENCRYPTION_ALGORITHM_ERROR|INVALID_LOCATION_CONSTRAINT|INVALID_OBJECT_STATE|INVALID_PART|
+        INVALID_PART_ORDER|INVALID_PAYER|INVALID_POLICY_DOCUMENT|INVALID_RANGE|INVALID_REQUEST|INVALID_SECURITY|
+        INVALID_SOAP_REQUEST|INVALID_STORAGE_CLASS|INVALID_TARGET_BUCKET_FOR_LOGGING|INVALID_TOKEN|INVALID_URI|
+        KEY_TOO_LONG_ERROR|MALFORMED_ACL_ERROR|MALFORMED_POST_REQUEST|MALFORMED_XML|MAX_MESSAGE_LENGTH_EXCEEDED|
+        MAX_POST_PRE_DATA_LENGTH_EXCEEDED_ERROR|METADATA_TOO_LARGE|METHOD_NOT_ALLOWED|MISSING_ATTACHMENT|
+        MISSING_CONTENT_LENGTH|MISSING_REQUEST_BODY_ERROR|MISSING_SECURITY_ELEMENT|MISSING_SECURITY_HEADER|
+        NO_LOGGING_STATUS_FOR_KEY|NO_SUCH_BUCKET|NO_SUCH_BUCKET_POLICY|NO_SUCH_KEY|NO_SUCH_LIFECYCLE_CONFIGURATION|
+        NO_SUCH_UPLOAD|NO_SUCH_VERSION|NOT_IMPLEMENTED|NOT_SIGNED_UP|OPERATION_ABORTED|PERMANENT_REDIRECT|
+        PRECONDITION_FAILED|REDIRECT|RESTORE_ALREADY_IN_PROGRESS|REQUEST_IS_NOT_MULTI_PART_CONTENT|REQUEST_TIMEOUT|
+        REQUEST_TIME_TOO_SKEWED|REQUEST_TORRENT_OF_BUCKET_ERROR|SERVER_SIDE_ENCRYPTION_CONFIGURATION_NOT_FOUND_ERROR|
+        SERVICE_UNAVAILABLE|SIGNATURE_DOES_NOT_MATCH|SLOW_DOWN|TEMPORARY_REDIRECT|TOKEN_REFRESH_REQUIRED|
+        TOO_MANY_BUCKETS|UNEXPECTED_CONTENT|UNRESOLVABLE_GRANT_BY_EMAIL_ADDRESS|USER_KEY_MUST_BE_SPECIFIED;
+
+public type BucketOperationError error<BUCKET_OPERATION_ERROR_REASONS, ErrorDetail>;
```

Comment:
This isn't the way we define `error`s. Combining all the error reasons into one does not serve the _intention of_ defining many error reasons.

The correct way is to define different constants for different error types, then define corresponding error types using the constants, and then (if needed) combine similar errors into union types using error type unions. Not by error reason unions.

This approach is also considered when the error types were designed, but we chose the aforementioned design over this, due to various usability issues.

Reply:
This is already there in the master branch [1]. I renamed the module name and did various changes to server error reasons. Due to this amazons3_errors.bal file is added as a new file.

[1] https://github.com/ballerina-platform/module-amazons3/blob/master/src/amazons3/amazons3_errors.bal#L145

Reply:
Agree with @ThisaruGuruge. We need to correct this. It should be something like https://github.com/ballerina-platform/ballerina-lang/blob/master/stdlib/grpc/src/main/ballerina/src/grpc/grpc_errors.bal
@praveennadarajah Shall we correct it now

Reply:
+1 Better we change it now. Since we are doing a new release on this.

Reply:
Fixed.